bee-tangle/src/tangle.rs

```diff
@@ -14,46 +14,112 @@ use crate::{vertex::Vertex, TransactionRef as TxRef};
 use dashmap::{mapref::entry::Entry, DashMap, DashSet};
+use log::info;
+use lru::LruCache;
+
+use std::{
+    collections::HashSet,
+    fmt::Debug,
+    marker::PhantomData,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        RwLock,
+    },
+};
+
+const CACHE_LEN: usize = 65536;
+
+/// A trait used to provide hooks for a tangle. The tangle acts as an in-memory cache and will use hooks to extend its
+/// effective volume. When an entry doesn't exist in the tangle cache and needs fetching, or when an entry gets
+/// inserted, the tangle will call out to the hooks in order to fulfil fulfil these actions.
+#[async_trait]
+pub trait Hooks<T> {
+    /// An error generated by these hooks.
+    type Error: Debug;
+
+    /// Fetch a transaction from some external storage medium.
+    async fn get(&self, hash: &Hash) -> Result<(Tx, T), Self::Error>;
+    /// Insert a transaction into some external storage medium.
+    async fn insert(&self, hash: Hash, tx: Tx, metadata: T) -> Result<(), Self::Error>;
+}
+
+/// Phoney default hooks that do nothing.
+pub struct NullHooks<T>(PhantomData<T>);
+
+impl<T> Default for NullHooks<T> {
+    fn default() -> Self {
+        Self(PhantomData)
+    }
+}
+
+#[async_trait]
+impl<T: Send + Sync> Hooks<T> for NullHooks<T> {
+    type Error = ();
 
-use std::collections::HashSet;
+    async fn get(&self, _hash: &Hash) -> Result<(Tx, T), Self::Error> {
+        Err(())
+    }
+
+    async fn insert(&self, _hash: Hash, _tx: Tx, _metadata: T) -> Result<(), Self::Error> {
+        Ok(())
+    }
+}
 
 /// A foundational, thread-safe graph datastructure to represent the IOTA Tangle.
-pub struct Tangle<T>
+pub struct Tangle<T, H = NullHooks<T>>
 where
-    T: Clone + Copy,
+    T: Clone,
 {
     pub(crate) vertices: DashMap<Hash, Vertex<T>>,
     pub(crate) children: DashMap<Hash, HashSet<Hash>>,
     pub(crate) tips: DashSet<Hash>,
-    // TODO: PriorityQueue with customizable priority for implementing cache eviction strategy
+
+    pub(crate) cache_counter: AtomicU64,
+    pub(crate) cache_queue: RwLock<LruCache<Hash, u64>>,
+
+    pub(crate) hooks: H,
 }
 
-impl<T> Default for Tangle<T>
+impl<T, H: Hooks<T>> Default for Tangle<T, H>
 where
-    T: Clone + Copy,
+    T: Clone,
+    H: Default,
 {
     fn default() -> Self {
+        Self::new(H::default())
+    }
+}
+
+impl<T, H: Hooks<T>> Tangle<T, H>
+where
+    T: Clone,
+{
+    /// Creates a new Tangle.
+    pub fn new(hooks: H) -> Self {
         Self {
             vertices: DashMap::new(),
             children: DashMap::new(),
             tips: DashSet::new(),
+
+            cache_counter: AtomicU64::new(0),
+            cache_queue: RwLock::new(LruCache::new(CACHE_LEN + 1)),
+
+            hooks,
         }
     }
-}
 
-impl<T> Tangle<T>
-where
-    T: Clone + Copy,
-{
-    /// Creates a new Tangle.
-    pub fn new() -> Self {
-        Self::default()
+    /// Create a new tangle with the given capacity.
+    pub fn with_capacity(self, cap: usize) -> Self {
```

Comment:
If not already the case, we should have a configuration parameter for the capacity and call this function instead of default to create the tangle.

Reply:
The problem with this is that it becomes exponential: you might also want to specify the hooks parameter (or not) in which case you end up needing to have 4 different constructor (if one includes `Default::default`). A builder method condenses this problem down into something simpler.

Reply:
I don't really mind how this is constructed, but the cache size should definitely be configurable to adapt to different servers.

Reply:
We could probably make this part of the tangle worker config in a later PR